integration/container_update_resources_test.go

```diff
@@ -39,7 +44,164 @@ func checkMemoryLimit(t *testing.T, spec *runtimespec.Spec, memLimit int64) {
+	if err != nil {
+		t.Fatal(err)
+	}
+	return stat.Memory.HierarchicalSwapLimit
+}
+
+func getCgroupMemoryLimitForTask(t *testing.T, task containerd.Task) uint64 {
+	if cgroups.Mode() == cgroups.Unified {
+		groupPath, err := cgroupsv2.PidGroupPath(int(task.Pid()))
+		if err != nil {
+			t.Fatal(err)
+		}
+		cgroup2, err := cgroupsv2.LoadManager("/sys/fs/cgroup", groupPath)
+		if err != nil {
+			t.Fatal(err)
+		}
+		stat, err := cgroup2.Stat()
+		if err != nil {
+			t.Fatal(err)
+		}
+		return stat.Memory.UsageLimit
+	}
+
+	cgroup, err := cgroups.Load(cgroups.V1, cgroups.PidPath(int(task.Pid())))
+	if err != nil {
+		t.Fatal(err)
+	}
+	stat, err := cgroup.Stat(cgroups.IgnoreNotExist)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return stat.Memory.Usage.Limit
+}
+
+func isSwapLikelyEnabled() bool {
+	// Check whether swap is enabled.
+	swapFile := "/proc/swaps"
+	swapData, err := ioutil.ReadFile(swapFile)
+	if err != nil {
+		// We can't read the file or it doesn't exist, assume we don't have swap.
+		return false
+	}
+
+	swapData = bytes.TrimSpace(swapData) // extra trailing \n
+	swapLines := strings.Split(string(swapData), "\n")
+	// If there is more than one line (table headers) in /proc/swaps, swap is enabled
+	return len(swapLines) > 1
+}
+
+func TestUpdateContainerResources_MemorySwap(t *testing.T) {
+	if !isSwapLikelyEnabled() {
+		t.Skipf("Swap is not enabled, or /proc/swaps is not readable. Swap is required for this test")
+		return
+	}
+
+	t.Log("Create a sandbox")
+	sb, sbConfig := PodSandboxConfigWithCleanup(t, "sandbox", "update-container-swap-resources")
+
+	EnsureImageExists(t, pauseImage)
+
+	memoryLimit := int64(128 * 1024 * 1024)
+	baseSwapLimit := int64(200 * 1024 * 1024)
+	increasedSwapLimit := int64(256 * 1024 * 1024)
+
+	t.Log("Create a container with memory limit but no swap")
+	cnConfig := ContainerConfig(
+		"container",
+		pauseImage,
+		WithResources(&runtime.LinuxContainerResources{
+			MemoryLimitInBytes:     memoryLimit,
+			MemorySwapLimitInBytes: baseSwapLimit,
+		}),
+	)
+	cn, err := runtimeService.CreateContainer(sb, cnConfig, sbConfig)
+	require.NoError(t, err)
+
+	t.Log("Check memory limit in container OCI spec")
+	container, err := containerdClient.LoadContainer(context.Background(), cn)
+	require.NoError(t, err)
+	spec, err := container.Spec(context.Background())
+	require.NoError(t, err)
+	checkMemoryLimit(t, spec, memoryLimit)
+	checkMemorySwapLimit(t, spec, &baseSwapLimit)
+
+	t.Log("Update container swap limit after created")
+	err = runtimeService.UpdateContainerResources(cn, &runtime.LinuxContainerResources{
+		MemorySwapLimitInBytes: baseSwapLimit,
+	}, nil)
+	require.NoError(t, err)
+
+	t.Log("Check memory limit in container OCI spec")
+	spec, err = container.Spec(context.Background())
+	require.NoError(t, err)
+	sw1 := baseSwapLimit
+	checkMemorySwapLimit(t, spec, &sw1)
+
+	t.Log("Start the container")
+	require.NoError(t, runtimeService.StartContainer(cn))
```

Comment:
```
=== RUN   TestUpdateContainerResources_MemorySwap
    container_update_resources_test.go:136: Create a sandbox
    common.go:115: Image "k8s.gcr.io/pause:3.6" already exists, not pulling.
    container_update_resources_test.go:145: Create a container with memory limit but no swap
    container_update_resources_test.go:157: Check memory limit in container OCI spec
    container_update_resources_test.go:165: Update container swap limit after created
    container_update_resources_test.go:171: Check memory limit in container OCI spec
    container_update_resources_test.go:177: Start the container
E1202 20:57:26.825490   29449 remote_runtime.go:265] StartContainer "fecc964cf066cfe7b7d08f63267042ffb7c30321c5858a9cc1d38eb5d8aa669a" from runtime service failed: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: process_linux.go:508: setting cgroup config for procHooks process caused: open /sys/fs/cgroup/memory/k8s.io/fecc964cf066cfe7b7d08f63267042ffb7c30321c5858a9cc1d38eb5d8aa669a/memory.memsw.limit_in_bytes: no such file or directory: unknown
    container_update_resources_test.go:178: 
        	Error Trace:	container_update_resources_test.go:178
        	Error:      	Received unexpected error:
        	            	rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: process_linux.go:508: setting cgroup config for procHooks process caused: open /sys/fs/cgroup/memory/k8s.io/fecc964cf066cfe7b7d08f63267042ffb7c30321c5858a9cc1d38eb5d8aa669a/memory.memsw.limit_in_bytes: no such file or directory: unknown
        	Test:       	TestUpdateContainerResources_MemorySwap
--- FAIL: TestUpdateContainerResources_MemorySwap (2.10s)
```

Reply:
So I just tested this and updated tests for cgroupsv1 on fedora 35 (based on the vagrant box in the project repo) - and didn't run into this (it just failed as expected bc of memsw vs swap differences).

Do you have any pointers for reproing?

Reply:
I think I've figured it out - It looks like Ubuntu has cgroup swap accounting disabled by default. This means we end up with no `memory.memsw.limit_in_bytes` file being created (and no way to control or report swap).

Because we're using GitHub Actions I'm not sure we can reconfigure grub to add `swapaccount=1` (or if that would be desired).

This means at the very least I need to extend swap detection to check swap accounting and skip the test on those nodes.

As a follow up we can introduce containerd-swap environments in k8s test-infra to cover the integration path, but I'm not sure how to make these tests run on GHA before they add non-Ubuntu envs.

Reply:
Updated to skip when swap accounting isn't possible - Mostly need guidance on how we would want to verify that going forwards.